crates/nu-plugin/src/plugin/mod.rs

```diff
@@ -265,16 +269,23 @@ pub trait Plugin {
 /// #         -> Result<Value, LabeledError> {todo!();}
 /// # }
 /// fn main() {
-///    serve_plugin(&mut MyPlugin::new(), MsgPackSerializer)
+///    serve_plugin(&mut MyPlugin::new(), EncodingType::MsgPack)
 /// }
 /// ```
 ///
 /// The object that is expected to be received by nushell is the `PluginResponse` struct.
 /// The `serve_plugin` function should ensure that it is encoded correctly and sent
 /// to StdOut for nushell to decode and and present its result.
-pub fn serve_plugin(plugin: &mut impl Plugin, encoder: impl PluginEncoder) {
+pub fn serve_plugin(plugin: &mut impl Plugin, encoder_format: EncodingType) {
+    match encoder_format {
+        EncodingType::Json => serve_plugin_with_format::<JsonSerializer>(plugin),
+        EncodingType::MsgPack => serve_plugin_with_format::<MsgPackSerializer>(plugin),
+    }
+}
```

Comment:
Mhh, I think there is a clear potential downside to the approach of using an enum here.

On the crate boundary this function will be compiled momorphizing `serve_plugin_with_format::<T: PluginEncoder>` for all options and using all variants of `serve_plugin_with_format` in the match, even though a plugin implementation using `nu-plugin` will typically only use one `PluginEncoder`/`EncodingType` variant. To get rid of the result of all this codegen (the serdes are big!) the dead-code elimination needs to constant fold across the crate boundary, which requires goodwill by LLVM or succesfull LTO. 

I would like to see the before and after comparison with `cargo bloat`/`cargo build --timings` for a plugin.
On the nushell side itself we won't be able to get rid of having the serde code for all formats we support, but we should be nice to the plugin authors in that regard.

Reply:
Thank you for the comment.

Just to make sure I understand your concern: is your objection that by referencing both serializers, plugins will bundle both serialization libraries thereby increasing the compiled binary size of the plugin? Or are there other concerns with monomophizing `serve_plugin_with_format for` both encoders?

Reply:
Yes the direct concern is just code/binary size and any downstream effects from that.

This adds close to 0.8 MB or 15% to `nu_plugin_formats` 

## Output from `cargo bloat -p nu_plugin_formats --release`

### This branch
Total: 6.1 MiB
`.text`: 2.7 MiB
<details>
   <summary>full output (function level breakdown)</summary>

```
    Analyzing target/release/nu_plugin_formats

 File  .text    Size        Crate Name
 0.6%   1.3% 35.9KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.3%   0.8% 21.7KiB          std addr2line::ResDwarf<R>::parse
 0.3%   0.7% 18.7KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.3%   0.6% 16.1KiB    nu_plugin <nu_protocol::shell_error::_::<impl serde::de::Deserialize for nu_protocol::shell_error::ShellError>::deserialize::__FieldVisitor as serde::de::Visitor>::visit_bytes
 0.2%   0.5% 15.0KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.2%   0.5% 12.9KiB        regex regex::exec::ExecBuilder::build
 0.2%   0.4% 10.6KiB          std addr2line::ResUnit<R>::parse_lines
 0.2%   0.4%  9.9KiB  serde_json? <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
 0.1%   0.3%  9.2KiB    nu_plugin nu_plugin::plugin::serve_plugin
 0.1%   0.3%  9.0KiB   eml_parser eml_parser::parser::EmlParser::parse
 0.1%   0.3%  8.2KiB  nu_protocol <alloc::vec::Vec<T,A> as core::clone::Clone>::clone
 0.1%   0.3%  8.1KiB          std miniz_oxide::inflate::core::decompress
 0.1%   0.3%  7.6KiB regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.1%   0.2%  6.8KiB          std <T as alloc::alloc::WriteCloneIntoRaw>::write_clone_into_raw
 0.1%   0.2%  6.6KiB aho_corasick aho_corasick::automaton::try_find_fwd
 0.1%   0.2%  6.6KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  6.3KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  6.2KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  5.9KiB    nu_plugin nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.2%  5.9KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
40.5%  91.7%  2.5MiB              And 8252 smaller methods. Use -n N to show more.
44.2% 100.0%  2.7MiB              .text section size, the file size is 6.1MiB
```
</details>

### `main` @ `a0dafcb0f` (effectively the base commit, deviation by one commit in a separate crate)

Total: 5.3 MiB
`.text`: 2.3 MiB
<details>
   <summary>full output (function level breakdown)</summary>
```
    Analyzing target/release/nu_plugin_formats

 File  .text    Size        Crate Name
 0.7%   1.5% 35.9KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.4%   0.9% 21.7KiB          std addr2line::ResDwarf<R>::parse
 0.3%   0.8% 18.7KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.3%   0.7% 16.1KiB    nu_plugin <nu_protocol::shell_error::_::<impl serde::de::Deserialize for nu_protocol::shell_error::ShellError>::deserialize::__FieldVisitor as serde::de::Visitor>::visit_bytes
 0.3%   0.6% 15.0KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.2%   0.6% 12.9KiB        regex regex::exec::ExecBuilder::build
 0.2%   0.5% 10.6KiB          std addr2line::ResUnit<R>::parse_lines
 0.2%   0.4%  9.0KiB   eml_parser eml_parser::parser::EmlParser::parse
 0.2%   0.4%  8.2KiB  nu_protocol <alloc::vec::Vec<T,A> as core::clone::Clone>::clone
 0.1%   0.3%  8.1KiB          std miniz_oxide::inflate::core::decompress
 0.1%   0.3%  7.6KiB regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.1%   0.3%  6.8KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  6.8KiB          std <T as alloc::alloc::WriteCloneIntoRaw>::write_clone_into_raw
 0.1%   0.3%  6.6KiB aho_corasick aho_corasick::automaton::try_find_fwd
 0.1%   0.3%  6.3KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  6.2KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  5.9KiB    nu_plugin nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.3%  5.9KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  5.9KiB  nu_protocol nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.3%  5.9KiB  nu_protocol <nu_protocol::shell_error::ShellError as core::fmt::Debug>::fmt
38.6%  90.3%  2.0MiB              And 7000 smaller methods. Use -n N to show more.
42.8% 100.0%  2.3MiB              .text section size, the file size is 5.3MiB
```
</details>